### ISSUE TITLE
Despite --skip-editable, --strict still fails due to editable packages

### ISSUE THREAD

**Issue by marcinbarczynski** [pypa/pip-audit#664]
## Bug description

According to the help message, 

>   --skip-editable       don't audit packages that are marked as editable (default: False)

I expected that `pip-audit --skip-editable --strict` would not consider editable packages at all.
However, `pip-audit` still fails with "distribution marked as editable" when it encounters an editable package.
That makes the `--strict` option of limited use.

## Platform information

* OS name and version:
* `pip-audit` version (`pip-audit -V`): pip-audit 2.6.1
* Python version (`python -V` or `python3 -V`): Python 3.11.4
* `pip` version (`pip -V` or `pip3 -V`): pip 23.2.1

**Comment by woodruffw** [comment]
Thanks for the report @marcinbarczynski. This sounds like a bug on our part.

Would you be able to share a small reproducer testcase?

**Comment by woodruffw** [comment]
In particular: are you seeing this in requirements (`-r`) mode, or in the default mode?

**Comment by woodruffw** [comment]
Making sure I understand: is the distinction you're drawing between "expected unaudited" dependencies and "unexpected unaudited" ones?

> With the current semantics, I see no reason for passing both --skip-editable and --strict as the tool fails anyway.

Sorry, I'm not following this: locally, each of the following does different things for me:

* `pip-audit`: don't skip editable dependencies, allow skips
* `pip-audit --strict`: don't skip editable dependencies, treat skips as failures
* `pip-audit --skip-editable`: skip editable dependencies, allow skips
* `pip-audit --skip-editable --strict`: skip editable dependencies and treat those skips as failures

Of these, the only one to fail for me locally is the last one (meaning exit with an `ERROR` on stderr, not just a nonzero exit code indicating audit status).

To your point: that last invocation is arguably less useful than the other three, since it's essentially saying "always fail if there's anything editable," which may or may not be useful depending on how you're using `pip-audit` (some people use it in CI where this wouldn't matter, others use it as a deployment check where editable installs would be a red flag). 

TL;DR: I'm not sure what/if we can change here: `--strict`'s behavior is defined to mean "fail on *any* skip", so we can't easily change that without breaking the existing behavioral contract. I suppose we could add another flag to special-case editable skips, but I'm wary of doing so (since it adds another configuration point/another state that needs to be fanned out across the various sources we support). 

This might be a good use case for `pip-audit --format=json --skip-editable` and then performing your own post-processing (i.e., checking all of the skips, and failing if any of them are ones you don't expect). Do you think that would work for your use case?

**Comment by marcinbarczynski** [comment]
> Making sure I understand: is the distinction you're drawing between "expected unaudited" dependencies and "unexpected unaudited" ones?

More or less. In my use case, I expect that editable packages are not auditable and want to skip them. OTOH, I would like to have a simple method to determine if something went wrong. An exit code is much easier than parsing json.

>> With the current semantics, I see no reason for passing both --skip-editable and --strict as the tool fails anyway.

> Sorry, I'm not following this: locally, each of the following does different things for me:

In my view, the only difference between `pip-audit --skip-editable --strict` and `pip-audit --strict` is that they fail with different error messages: "distribution marked as editable" vs. "Dependency not found on PyPI and could not be audited".

After reading the help message,
```
  --skip-editable       don't audit packages that are marked as editable (default: False)
  -S, --strict          fail the entire audit if dependency collection fails on any dependency (default: False)
```
I expected that "don't audit package" means that no dependency collection occurs.

**Comment by woodruffw** [comment]
> What is a use case for that?

They signal different error states: a user might have a distribution that *is* on PyPI that's also marked as editable, in which case `pip-audit --strict` will work while `pip-audit --skip-editable --strict` will fail. 

Your use case is one we'd ideally handle (where editable packages are not auditable), but we can't do it by incompatibly changing another use case that we already support, unfortunately. I'll think some more about how we can accommodate this (ideally without adding more CLI flags, since we're trying to keep them to a minimum).

> I expected that "don't audit package" means that no dependency collection occurs.

This is an understandable interpretation, but it isn't what the help message means to say: dependency resolution *always* happens unless the user passes `--no-deps` or `--require-hashes` (or enters those mode(s) by providing a hashed input).

(However, in the context where editable packages occur, this is a moot point: editable packages only appear in environments, and environments don't perform any dependency resolution since they already list all of their dependencies.)

**Comment by marcinbarczynski** [comment]
Thanks for the explanation. I've already worked around the current semantics for my use case.
